arch/arm/core/cortex_m/mpu/nxp_mpu.c

```diff
@@ -194,64 +243,42 @@ void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
 	SYS_LOG_DBG("Region info: 0x%x 0x%x", base, size);
 	u32_t region_index = _get_region_index_by_type(type);
 	u32_t region_attr = _get_region_attr_by_type(type);
-	u32_t last_region = _get_num_regions() - 1;
 
-	/*
-	 * The NXP MPU manages the permissions of the overlapping regions
-	 * doing the logic OR in between them, hence they can't be used
-	 * for stack/stack guard protection. For this reason the last region of
-	 * the MPU will be reserved.
-	 *
-	 * A consequence of this is that the SRAM is splitted in different
-	 * regions. In example if THREAD_STACK_GUARD_REGION is selected:
-	 * - SRAM before THREAD_STACK_GUARD_REGION: RW
-	 * - SRAM THREAD_STACK_GUARD_REGION: RO
-	 * - SRAM after THREAD_STACK_GUARD_REGION: RW
-	 */
 	/* NXP MPU supports up to 16 Regions */
 	if (region_index > _get_num_regions() - 2) {
 		return;
 	}
 
-	/* Configure SRAM_0 region */
-	/*
-	 * The mpu_config.sram_region contains the index of the region in
-	 * the mpu_config.mpu_regions array but the region 0 on the NXP MPU
-	 * is the background region, so on this MPU the regions are mapped
-	 * starting from 1, hence the mpu_config.sram_region on the data
-	 * structure is mapped on the mpu_config.sram_region + 1 region of
-	 * the MPU.
-	 */
-	_region_init(mpu_config.sram_region,
-		     mpu_config.mpu_regions[mpu_config.sram_region].base,
-		     ENDADDR_ROUND(base),
-		     mpu_config.mpu_regions[mpu_config.sram_region].attr);
-
-	switch (type) {
-	case THREAD_STACK_REGION:
-		break;
-	case THREAD_STACK_GUARD_REGION:
-		_region_init(region_index, base,
-			     ENDADDR_ROUND(base + STACK_GUARD_REGION_SIZE),
-			     region_attr);
-		break;
-	default:
-		break;
-	}
-
-	/* Configure SRAM_1 region */
-	_region_init(last_region,
-		     base + STACK_GUARD_REGION_SIZE,
-		     ENDADDR_ROUND(
-			    mpu_config.mpu_regions[mpu_config.sram_region].end),
-		     mpu_config.mpu_regions[mpu_config.sram_region].attr);
-
+	_region_init(region_index, base,
+		     ENDADDR_ROUND(base + size),
+		     region_attr);
+	if (type == THREAD_STACK_GUARD_REGION)
+		nxp_mpu_setup_sram_region(base, size);
 }
 
 #if defined(CONFIG_USERSPACE)
 void arm_core_mpu_configure_user_context(struct k_thread *thread)
 {
-	return 0;
+	u32_t base = (u32_t)thread->stack_obj + MPU_GUARD_ALIGN_AND_SIZE;
+	u32_t size = thread->stack_info.size;
+	u32_t index = _get_region_index_by_type(THREAD_STACK_USER_REGION);
+	u32_t region_attr = _get_region_attr_by_type(THREAD_STACK_USER_REGION);
+
+	/* configure stack */
+	_region_init(index, base, ENDADDR_ROUND(base + size), region_attr);
+
+	/* configure app data portion */
+	index = _get_region_index_by_type(THREAD_APP_DATA_REGION);
+	region_attr = _get_region_attr_by_type(THREAD_APP_DATA_REGION);
+	base = (u32_t)&__app_ram_start;
+	size = (u32_t)&__app_ram_end - (u32_t)&__app_ram_start;
+
+	/* set up app data region if exists, otherwise disable */
+	if (size > 0)
```

Comment:
zephyr coding style insists on braces for all if clauses even if they are one-liners